### ISSUE TITLE
Init script reports complete before sssd is actually working

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#1361]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/1357
- Created at 2012-05-29 23:17:23 by [tljohnsn](https://pagure.io/user/tljohnsn)
- Closed as Fixed
- Assigned to [pbrezina](https://pagure.io/user/pbrezina)
---

I installed a rhel 6.2 system with the sshd Privilege-separated user account only in active directory and not in the system /etc/passwd file.  Under these conditions sshd which is supposed to run directly after sssd will not start.

Instead the following error is returned:
[  OK  ] sssd: [  OK  ]
Starting sshd: Privilege separation user sshd does not exist
[FAILED]

The system sssd version is sssd-1.5.1-66.el6_2.3.x86_64

### Comments

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-05-30 00:24:51](https://pagure.io/SSSD/sssd/issue/1357#234316)**

How did you configure SSSD?

Can you please add the sanitized sssd.conf file and turn on debugging for the sssd to debug_level=9 and attach a sanitized log that reproduces this issue?

Also nsswitch.conf would be handy.

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-05-30 08:52:05](https://pagure.io/SSSD/sssd/issue/1357#234317)**

I suspect this could have the same root cause as https://fedorahosted.org/sssd/ticket/1303 That is, when the request for the ssh user comes in, the SSSD NSS provider is not up yet.

Does sshd start afterwards manually, or if you added some artificial delay in the init script?

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-05-30 14:11:28](https://pagure.io/SSSD/sssd/issue/1357#234318)**

Well, I think this is a slightly different issue. Yes, we're taking more time than we'd like to start up, but the bug here is that 'service sssd start' returns success before the SSSD is actually fully started.

Essentially, the init script will always return success as soon as the monitor starts and its pidfile is created. Maybe we need to defer the creation of the pidfile until the responders have been successfully started. This will most likely slow down startup, since it can take a couple seconds for slow domains to load.

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-05-30 17:53:57](https://pagure.io/SSSD/sssd/issue/1357#234319)**

attachment
[sssd.log](/SSSD/sssd/issue/raw/files/400cff717ed36f44c13ad06b90057dc5c03d7946a17246f045c392220f8ad6a0-sssd.log)

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-05-30 17:54:08](https://pagure.io/SSSD/sssd/issue/1357#234320)**

attachment
[sssd.conf](/SSSD/sssd/issue/raw/files/92b2a1aca31672cfe4070e2815ef15d8e116f972612d2f40991084d44a74dcc2-sssd.conf)

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-05-30 17:54:28](https://pagure.io/SSSD/sssd/issue/1357#234321)**

attachment
[nsswitch.conf](/SSSD/sssd/issue/raw/files/806da1a4aea48a82b5c2dcdd8053dc4fe974c085f07bf6754dc1a45b4faeede9-nsswitch.conf)

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-05-30 17:57:51](https://pagure.io/SSSD/sssd/issue/1357#234322)**

I have attached the requested sssd configuration with debug_level=9, as well as the sssd.log file and nsswitch.conf.  Please let me know if you require any of the other sssd log files.

I have worked around the problem by creating another init script that runs between sssd and sshd.  That script does nothing but sleep five seconds.  With that hack in place, or by starting sshd manually on the console, things will work properly.

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-06-07 09:11:40](https://pagure.io/SSSD/sssd/issue/1357#234323)**

Replying to [comment:3 sgallagh]:
> Well, I think this is a slightly different issue. Yes, we're taking more time than we'd like to start up, but the bug here is that 'service sssd start' returns success before the SSSD is actually fully started.
> 
> Essentially, the init script will always return success as soon as the monitor starts and its pidfile is created. Maybe we need to defer the creation of the pidfile until the responders have been successfully started. This will most likely slow down startup, since it can take a couple seconds for slow domains to load.

This is starting to sound like a good idea, I got another report from a different user that has the same problem with autofs.

Maybe we could only wait with reporting that startup has ended until the responders are up, the responders can return most of the data from cache until the providers are started.

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-06-07 13:30:24](https://pagure.io/SSSD/sssd/issue/1357#234324)**

Well, in most cases this is actually no different from waiting for the providers to start up.

In the monitor code, we start the providers first and then wait for them to complete[*] before starting the responders.

[*] Well, if the providers do not start up within five seconds, we force the providers to start up anyway.

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-06-07 15:18:08](https://pagure.io/SSSD/sssd/issue/1357#234325)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.9.0

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-06-07 15:18:44](https://pagure.io/SSSD/sssd/issue/1357#234326)**

Ticket has been cloned to Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=829740

rhbz:  => [https://bugzilla.redhat.com/show_bug.cgi?id=829740 829740]

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-09-10 16:32:33](https://pagure.io/SSSD/sssd/issue/1357#234327)**

Fields changed

owner: somebody => okos

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-09-24 17:04:37](https://pagure.io/SSSD/sssd/issue/1357#234328)**

Fields changed

milestone: SSSD 1.9.0 => SSSD 1.9.1

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-09-27 21:07:08](https://pagure.io/SSSD/sssd/issue/1357#234329)**

Fields changed

owner: okos => pbrezina
status: new => assigned

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-09-27 21:16:54](https://pagure.io/SSSD/sssd/issue/1357#234330)**

Fields changed

patch: 0 => 1

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-10-02 16:25:22](https://pagure.io/SSSD/sssd/issue/1357#234331)**

master:

- fa893b2796b002f709e9416f134bc8df8c08cf8d
- f3c6d7d83f1906369fc64a1480d1c014f693b0cd

resolution:  => fixed
status: assigned => closed

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-10-15 17:23:42](https://pagure.io/SSSD/sssd/issue/1357#234332)**

Looks like there is still an issue:

    [root@ipa17-devel ~]# /etc/init.d/sssd stop && rm -f /var/lib/sss/db/*.ldb /var/lib/sss/mc/* /var/log/sssd/* && /etc/init.d/sssd start && date && id admin; sleep 3 && date && id admin
    Stopping sssd (via systemctl):                             [  OK  ]
    Starting sssd (via systemctl):                             [  OK  ]
    Mo 15. Okt 17:12:30 CEST 2012
    id: admin: no such user
    Mo 15. Okt 17:12:33 CEST 2012
    uid=1662200000(admin) gid=1662200000(admins) Gruppen=1662200000(admins)

According to the logs the nss responder is not started at the time of the first request:

    (Mon Oct 15 17:12:31 2012) [sssd[nss]] [server_setup] (0x0400): CONFDB: /var/lib/sss/db/config.ldb

and only the second request is recorded in the log files.

milestone: SSSD 1.9.1 => NEEDS_TRIAGE
resolution: fixed => 
status: closed => reopened

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-10-15 18:03:13](https://pagure.io/SSSD/sssd/issue/1357#234333)**

Can you paste sssd.log and configuration please?

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-10-16 13:48:50](https://pagure.io/SSSD/sssd/issue/1357#234334)**

It looks like systemd ignores our pid file. In the time when systemd reports OK, the pid file does not exist.

    /etc/init.d/sssd stop && rm -f /var/lib/sss/db/*.ldb /var/lib/sss/mc/* /var/log/sssd/* && /etc/init.d/sssd start && date && cat /var/run/sssd.pid ; id user-1; sleep 3 && date && cat /var/run/sssd.pid ; id user-1
    Stopping sssd (via systemctl):                             [  OK  ]
    Starting sssd (via systemctl):                             [  OK  ]
    Tue Oct 16 13:47:44 CEST 2012
    cat: /var/run/sssd.pid: No such file or directory
    id: user-1: No such user
    Tue Oct 16 13:47:47 CEST 2012
    31118
    uid=10001(user-1) gid=10001 groups=10001

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-10-16 14:12:19](https://pagure.io/SSSD/sssd/issue/1357#234335)**

We use "forking" type in unit file. According to man systemd.service, systemd does not monitor existence of the pid file, it just reads it to know what is the main process. Systemd returns OK when the original process exits (after fork).

We have two options here:

- postpone exiting after fork
- become another type of service (probably notify)

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-10-16 14:31:25](https://pagure.io/SSSD/sssd/issue/1357#234336)**

Can you check out if sysv behaves sanely in this respect?

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-10-18 15:02:43](https://pagure.io/SSSD/sssd/issue/1357#234337)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.9.3

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-11-06 12:08:57](https://pagure.io/SSSD/sssd/issue/1357#234338)**

Fixed (again) in sssd-1-9:

- 3a3e1a3 create pid file immediately after fork again
- d80485d exit original process after sssd is initialized
- 6ec69c2 make monitor_quit() usable outside signal handler
- 8b130e4 fix indendation, coding style and debug levels in server.c
- 3d73923 add SSSDBG_IMPORTANT_INFO macro

and master:

- d19c478 create pid file immediately after fork again
- 715e09e exit original process after sssd is initialized
- e02ec73 make monitor_quit() usable outside signal handler
- 53b475b fix indendation, coding style and debug levels in server.c
- fa3e287 add SSSDBG_IMPORTANT_INFO macro

design:  => 
design_review:  => 0
fedora_test_page:  =>

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2012-11-06 12:13:17](https://pagure.io/SSSD/sssd/issue/1357#234339)**

Fields changed

resolution:  => fixed
status: reopened => closed

---

**Comment from [tljohnsn](https://pagure.io/user/tljohnsn) at [2017-02-24 14:49:29](https://pagure.io/SSSD/sssd/issue/1357#234340)**

**Metadata Update from @tljohnsn**:
- Issue assigned to pbrezina
- Issue set to the milestone: SSSD 1.9.3
